internal/coreinternal/scraperinttest/scraperint.go

```diff
@@ -0,0 +1,179 @@
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/component/componenttest"
+	"go.opentelemetry.io/collector/consumer/consumertest"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/receiver"
+	"go.opentelemetry.io/collector/receiver/receivertest"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/golden"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/pdatatest/pmetrictest"
+)
+
+func EqualsLatestMetrics(expected pmetric.Metrics, sink *consumertest.MetricsSink, compareOpts []pmetrictest.CompareMetricsOption) func() bool {
+	return func() bool {
+		allMetrics := sink.AllMetrics()
+		return len(allMetrics) > 0 && nil == pmetrictest.CompareMetrics(expected, allMetrics[len(allMetrics)-1], compareOpts...)
+	}
+}
+
+func NewIntegrationTest(f receiver.Factory, cr testcontainers.ContainerRequest, opts ...TestOption) *IntegrationTest {
+	it := &IntegrationTest{
+		factory:          f,
+		containerRequest: cr,
+		customConfig:     func(component.Config, string, MappedPortFunc) {},
+		expectedFile:     filepath.Join("testdata", "integration", "expected.yaml"),
+		compareTimeout:   time.Minute,
+	}
+	for _, opt := range opts {
+		opt(it)
+	}
+	return it
+}
+
+type IntegrationTest struct {
+	containerRequest testcontainers.ContainerRequest
+
+	factory      receiver.Factory
+	customConfig customConfigFunc
+
+	expectedFile   string
+	compareOptions []pmetrictest.CompareMetricsOption
+	compareTimeout time.Duration
+}
+
+func (it *IntegrationTest) Run(t *testing.T) {
+	ctx := context.Background()
+
+	require.NoError(t, it.containerRequest.Validate())
+	container, err := testcontainers.GenericContainer(ctx,
+		testcontainers.GenericContainerRequest{
+			ContainerRequest: it.containerRequest,
+			Started:          true,
+		})
+	require.NoError(t, err)
+	defer func() {
+		require.NoError(t, container.Terminate(context.Background()))
+	}()
+
+	cfg := it.factory.CreateDefaultConfig()
+	host, mappedPort := containerInfo(ctx, t, container)
+	it.customConfig(cfg, host, mappedPort)
+
+	sink := new(consumertest.MetricsSink)
+	settings := receivertest.NewNopCreateSettings()
+
+	rcvr, err := it.factory.CreateMetricsReceiver(ctx, settings, cfg, sink)
+	require.NoError(t, err, "failed creating metrics receiver")
+	require.NoError(t, rcvr.Start(ctx, componenttest.NewNopHost()))
+	defer func() {
+		require.NoError(t, rcvr.Shutdown(ctx))
+	}()
+
+	expected, err := golden.ReadMetrics(it.expectedFile)
+	require.NoError(t, err)
+
+	// Defined outside of Eventually so it can be printed if the test fails
+	var validateErr error
+	defer func() {
+		if t.Failed() {
+			t.Error(validateErr.Error())
+		}
+	}()
+
+	require.Eventually(t,
+		func() bool {
+			allMetrics := sink.AllMetrics()
+			if len(allMetrics) == 0 {
+				return false
+			}
+			validateErr = pmetrictest.CompareMetrics(expected, allMetrics[len(allMetrics)-1], it.compareOptions...)
+			return validateErr == nil
```

Comment:
Could `EqualsLatestMetrics` be used instead here?

Reply:
`EqualsLatestMetrics` was a good first step toward extracting some code but I think it can be fully removed as soon as a few more uses are migrated to the overall framework. The similar usage here is actually a good example of why it is better as a part of another function - because we're able to capture and log the error returned by `pmetrictest`, while simultaneously using the result as a condition within the `Eventually` call.